Add initial docs for the analysis-report package (#21878)
Yoast/wordpress-seo: packages/analysis-report/README.md

# @yoast/analysis-report

React components for visualizing SEO and readability analysis results from YoastSEO.js.

## Overview

> This package provides React components to display the results of content analysis performed by the [@yoast/yoastseo](../yoastseo) package. It visualizes SEO scores, readability metrics, and improvement suggestions in an organized and user-friendly way.

Reviewer: ```suggestion
This package provides React components to display the results of content analysis performed by the `yoastseo` package ([GitHub](../yoastseo), [NPM](https://www.npmjs.com/package/yoastseo)). It visualizes SEO scores, readability metrics, and improvement suggestions in an organized and user-friendly way.
```

## Installation

```bash
yarn add @yoast/analysis-report
```

Note: This package has peer dependencies that need to be installed:
- react
- react-dom
- styled-components

> - @wordpress/i18n

Reviewer: This bit can be removed in my opinion, they're not peer dependencies but actual dependencies.

      onMarkButtonClick={(id, marker) => {
        // Handle text marking
      }}
    />
  );
}
```

### AnalysisResult
Individual result item with score indicator and optional marking functionality:

#### Props
```typescript
interface AnalysisResultProps {
  // Content
  text: string;             // HTML string with analysis feedback
  score: number;            // 0-100 score
  rating: "good" | "OK" | "bad" | "feedback" | "upsell";
  id: string;              // Unique identifier
  
  // Marking functionality
  hasMarks: boolean;       // Whether this result supports text marking
  marker?: Marker;         // Marking instructions
  marksButtonStatus?: "enabled" | "disabled" | "hidden";
  
  // Edit functionality
  hasEditButton: boolean;  // Whether to show edit button
  editFieldName?: string;  // Field to focus when edit clicked
  
  // Styling
  bulletColor: string;     // Color for score indicator
  marksButtonClassName?: string;
  editButtonClassName?: string;
  
  // Accessibility
  ariaLabelMarks: string;
  ariaLabelEdit?: string;
  
  // State
  pressed: boolean;        // Whether marks are active
  suppressedText?: boolean;// Grey out the text
}
```

#### Features

> - Sanitizes HTML in text content

Reviewer: ```suggestion
- Sanitizes HTML in the analysis feedback
```

        // Handle button click
      }}
    />
  );
}
```

### SiteSEOReport  
Displays overall SEO score with a stacked progress bar:

#### Props
```typescript
interface SiteSEOReportProps {
  className?: string;
  seoAssessmentText: string;
  seoAssessmentItems: Array<{
    html: string;          // HTML content for item description
    value: number;         // Percentage value (0-100)
    color: string;         // Color for progress bar segment
  }>;
  barHeight?: string;      // Custom height for progress bar
}
```

#### Features
- Renders stacked progress bars
- Displays score breakdowns
- Supports custom styling and dimensions
- Handles responsive layouts

#### Example Usage
```jsx
import { SiteSEOReport } from "@yoast/analysis-report";

function Overview() {
  const items = [
    {
      value: 60,
      color: "#7ad03a",
      html: "60% of your pages have good SEO scores"
    },
    // ...more items
  ];

  return (
    <SiteSEOReport
      seoAssessmentText="Overall SEO score"
      seoAssessmentItems={items}
    />
  );
}
```

## Component Architecture

```mermaid
graph TD
    A[ContentAnalysis] --> B[AnalysisList]
    B --> C[AnalysisResult]
    C --> D[ScoreIcon]

>     C --> E[MarkerButtons]
>     C --> F[EditButton]

Reviewer: ```suggestion
    C --> E[Highlighting button]
    C --> F[Edit button (Premium)]
    C --> G[Yoast AI Optimize button (Premium)]
```

```

## Usage Outside of WordPress

When using this package outside of WordPress, keep in mind:

1. **Translations**: The package uses @wordpress/i18n for translations. You'll need to:

>    - Provide your own translation implementation

Reviewer: ```suggestion
   - Provide your own translation implementation when used outside of WordPress. The translations can be retrieved from https://translate.wordpress.org.
```

   - Mock the i18n functions in tests
   - Set up the translation infrastructure

> 2. **Styling**: The package uses styled-components and expects certain theme variables:

Reviewer: ```suggestion
2. **Styling**: The package uses `styled-components` and expects certain theme variables:
```

   - Supports keyboard navigation
   - Provides visual feedback on focus/hover

### Integration Example

```jsx
function SEOAnalysisWithMarking({ content }) {
  const [activeMarker, setActiveMarker] = useState("");
  const [marks, setMarks] = useState([]);

  const handleMarkClick = (id, marker) => {
    if (activeMarker === id) {
      // Remove marks
      setMarks([]);
      setActiveMarker("");
    } else {
      // Apply new marks
      const newMarks = typeof marker === "function" 
        ? marker() 
        : marker;
      setMarks(newMarks);
      setActiveMarker(id);
    }
  };

  const handleEditClick = (id, fieldName) => {
    const field = document.querySelector(`[name="${fieldName}"]`);
    if (field) {
      field.focus();
      field.scrollIntoView({ behavior: "smooth" });
    }
  };

  return (
    <>
      <ContentAnalysis
        {...results}
        marksButtonActivatedResult={activeMarker}
        onMarkButtonClick={handleMarkClick}
        onEditButtonClick={handleEditClick}
      />
      <TextEditor
        content={content}
        marks={marks}
        onMarkingRequired={handleMarkClick}
      />
    </>
  );
}
```

> ## Components Reference

Reviewer 2: This section is pretty much a duplicate of the `Core Components` section, except from the component descriptions that are slightly more detailed here.

}
```

#### Features
- Renders stacked progress bars
- Displays score breakdowns
- Supports custom styling and dimensions
- Handles responsive layouts

## Component Interactions

### Data Flow
```mermaid
sequenceDiagram
    participant P as Parent App
    participant C as ContentAnalysis
    participant A as AnalysisResult
    participant M as Marker System
    
    P->>C: Analysis Results
    C->>A: Individual Result
    A->>M: Marking Request
    M->>P: Mark Positions
    P->>M: Apply Marks
    M->>A: Update UI
```

### State Management

The components follow a unidirectional data flow:

1. **Parent Level**
   ```javascript
   {
     activeMarker: string | null;    // Currently active marker
     marks: Mark[];                  // Active text marks
     results: AnalysisResult[];      // Analysis results
   }
   ```

2. **ContentAnalysis Level**
   ```javascript
   {
     collapsibleStates: {           // Section collapse states
       [category: string]: boolean
     }
   }
   ```

3. **AnalysisResult Level**
   ```javascript
   {
     isHovered: boolean;            // UI hover state
     isPressed: boolean;            // Marker active state
   }

>    ```

Reviewer: I would remove these parts as they mostly repeat information from above.

Reviewer 2: I'd definitely keep the sequence diagram, that's nice to have. This last State Management section I don't really understand, but seems different from what I read before as well.
